src/core/Instrument.cpp

```diff
@@ -78,8 +82,83 @@ bool Instrument::isFromTrack( const Track * _track ) const
 	return( m_instrumentTrack == _track );
 }
 
+int count_zero_crossings(sampleFrame *buf,fpp_t start, fpp_t frames)
+{
+	// zero point crossing counts of all channels
+	int zero_crossings[DEFAULT_CHANNELS] = {0};
+	// maximum zero point crossing of all channels
+	int max_zc = 0;
+
+	// determine the zero point crossing counts
+	for (fpp_t f = start; f < frames; ++f)
+	{
+		for (ch_cnt_t ch=0; ch < DEFAULT_CHANNELS; ++ch)
+		{
+			// we don't want to count [-1, 0, 1] as two crossings
+			if ((buf[f-1][ch] <= 0.0 && buf[f][ch] > 0.0) ||
+					(buf[f-1][ch] >= 0.0 && buf[f][ch] < 0.0))
+			{
+				++zero_crossings[ch];
+				if (zero_crossings[ch] > max_zc)
+				{
+					max_zc = zero_crossings[ch];
+				}
+			}
+		}
+	}
+
+	return max_zc;
+}
+
+
+void Instrument::applyFadeIn(sampleFrame * buf, NotePlayHandle * n)
+{
+	f_cnt_t total = n->totalFramesPlayed();
+	if (total == 0)
+	{
+		const fpp_t frames = n->framesLeftForCurrentPeriod();
+
+		int max_zc = count_zero_crossings(buf, 1, frames);
```

Comment:
Following the PR's comments, I think the `1` is used in order to skip the first sample, which always produced a zero-crossing? If that's the case, I'd add a comment explaining why you used `1`. For us, it's clear, but a future reader might be confused.